packages/legend-query/src/stores/QueryBuilderMilestoningHelper.ts

```diff
@@ -0,0 +1,207 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { guaranteeType } from '@finos/legend-shared';
+import {
+  Class,
+  type PureModel,
+  type ValueSpecification,
+  DerivedProperty,
+  getMilestoneTemporalStereotype,
+  MILESTONING_STEREOTYPE,
+  type AbstractPropertyExpression,
+  Association,
+  INTERNAL__PropagatedValue,
+  PrimitiveInstanceValue,
+  PRIMITIVE_TYPE,
+  VariableExpression,
+} from '@finos/legend-graph';
+import type { QueryBuilderSetupState } from './QueryBuilderSetupState';
+import type { QueryBuilderState } from './QueryBuilderState';
+import type { QueryBuilderDerivedPropertyExpressionState } from './QueryBuilderPropertyEditorState';
+
+export const valueSpecifiation_isEqual = (
+  param1: ValueSpecification | undefined,
+  param2: ValueSpecification | undefined,
+): boolean => {
+  if (
+    param1 instanceof VariableExpression &&
+    param2 instanceof VariableExpression
+  ) {
+    return param1.name === param2.name;
+  } else if (
+    param1 instanceof PrimitiveInstanceValue &&
+    param2 instanceof PrimitiveInstanceValue
+  ) {
+    if (
+      param1.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE &&
+      param2.genericType.value.rawType.name === PRIMITIVE_TYPE.LATESTDATE
+    ) {
+      return true;
+    } else if (
+      param1.genericType.value.rawType.name ===
+      param2.genericType.value.rawType.name
+    ) {
+      return param1.values[0] === param2.values[0];
+    }
+    return false;
+  }
+  return false;
+};
+
+export const getSourceTemporalStereotype = (
+  property: DerivedProperty,
+  graph: PureModel,
+): MILESTONING_STEREOTYPE | undefined => {
+  const owner = property.owner;
+  if (owner instanceof Class) {
+    return getMilestoneTemporalStereotype(owner, graph);
+  } else if (owner instanceof Association) {
+    if (owner._generatedMilestonedProperties.length) {
+      const ownerClass =
+        owner._generatedMilestonedProperties[0]?.genericType.value.rawType;
+      return getMilestoneTemporalStereotype(
+        guaranteeType(ownerClass, Class),
+        graph,
+      );
+    }
+  }
+  return undefined;
+};
+
+export const checkForOverridingMilestoningQuery = (
+  derivedPropertyExpressionState: QueryBuilderDerivedPropertyExpressionState,
+  targetStereotype: MILESTONING_STEREOTYPE,
+): boolean => {
+  const sourceStereotype = getSourceTemporalStereotype(
+    derivedPropertyExpressionState.derivedProperty,
+    derivedPropertyExpressionState.queryBuilderState.graphManagerState.graph,
+  );
+  if (
+    sourceStereotype === MILESTONING_STEREOTYPE.PROCESSING_TEMPORAL &&
+    targetStereotype === MILESTONING_STEREOTYPE.BITEMPORAL &&
+    derivedPropertyExpressionState.parameterValues.length === 1
+  ) {
+    return true;
+  } else {
+    return false;
+  }
+};
+
+/**
+ * Checks if the parameter of the current milestoned property is same as `businessDate` or `processingDate`
```

Comment:
--> 
```ts
/**
 * Check if the parameter value of the milestoned property is
 * the same as those specified in global scope, so that we can
 * potentially replace them with propgated value.
 */
```